File: algorithms/16_kotlin/math/KalmanFilter.kt
package math

class KalmanFilter1D(var x: Double, var p: Double, var q: Double, var r: Double) {
    fun predict() { p += q }
    fun update(z: Double): Double {
        val k = p / (p + r)
        x += k * (z - x)
        p = (1.0 - k) * p
        return x
    }
}

fun main() {
    val kf = KalmanFilter1D(0.0, 1.0, 0.01, 0.1)
    for (z in doubleArrayOf(0.9, 1.1, 0.95, 1.05)) {
        kf.predict()
        kf.update(z)
    }
    assert(Math.abs(kf.x - 1.0) < 0.2)
    println("Kotlin Kalman Filter verified.")
}

File: algorithms/15_ruby/math/kalman_filter.rb
# 1D Kalman Filter in Ruby (Numerical Recipes 3rd Ed. Chapter 15)

class KalmanFilter1D
  attr_accessor :x, :p, :q, :r
  def initialize(x, p, q, r)
    @x = x; @p = p; @q = q; @r = r
  end

  def predict
    @p += @q
  end

  def update(z)
    k = @p / (@p + @r)
    @x += k * (z - @x)
    @p = (1.0 - k) * @p
    @x
  end
end

kf = KalmanFilter1D.new(0.0, 1.0, 0.01, 0.1)
[0.9, 1.1, 0.95, 1.05].each do |z|
  kf.predict
  kf.update(z)
end
raise "Failed" if (kf.x - 1.0).abs > 0.2
puts "Ruby Kalman Filter verified."
